sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,867 @@
+                lwmac_set_state(gnrc_netdev, LWMAC_TRANSMITTING);
+                return;
+            }
+            neighbour = _next_tx_neighbor(gnrc_netdev);
+        }
+
+        if (neighbour != NULL) {
+            /* if phase is unknown, send immediately. */
+            if (neighbour->phase > RTT_TICKS_TO_US(LWMAC_WAKEUP_INTERVAL_US)) {
+                gnrc_netdev->tx.current_neighbor = neighbour;
+                gnrc_netdev_lwmac_set_tx_continue(gnrc_netdev, false);
+                gnrc_netdev->tx.tx_burst_count = 0;
+                lwmac_set_state(gnrc_netdev, LWMAC_TRANSMITTING);
+                return;
+            }
+
+            /* Offset in microseconds when the earliest (phase) destination
+             * node wakes up that we have packets for. */
+            int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+            /* If there's not enough time to prepare a WR to catch the phase
+             * postpone to next interval */
+            if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+            }
+            time_until_tx -= LWMAC_WR_PREPARATION_US;
+
+            /* add a random time before goto TX, for avoiding one node for
+             * always holding the medium (if the receiver's phase is recorded earlier in this
+             * particular node) */
+            uint32_t random_backoff;
+            random_backoff = random_uint32_range(0, LWMAC_TIME_BETWEEN_WR_US);
+            time_until_tx = time_until_tx + random_backoff;
+
+            lwmac_set_timeout(gnrc_netdev, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+            /* Register neighbour to be the next */
+            gnrc_netdev->tx.current_neighbor = neighbour;
+
+            /* Stop dutycycling, we're preparing to send. This prevents the
+             * timeout arriving late, so that the destination phase would
+             * be missed. */
+            /* TODO: bad for power savings */
+            rtt_handler(LWMAC_EVENT_RTT_PAUSE, gnrc_netdev);
+        }
+    }
+    else {
+        if (lwmac_timeout_is_expired(gnrc_netdev, TIMEOUT_WAIT_FOR_DEST_WAKEUP)) {
+            LOG(LOG_DEBUG, "[lwmac] Got timeout for dest wakeup, ticks: %" PRIu32 "\n",
+                           rtt_get_counter());
+            gnrc_netdev_lwmac_set_tx_continue(gnrc_netdev, false);
+            gnrc_netdev->tx.tx_burst_count = 0;
+            lwmac_set_state(gnrc_netdev, LWMAC_TRANSMITTING);
+        }
+    }
+}
+
+static void _rx_management(gnrc_netdev_t *gnrc_netdev)
+{
+    lwmac_rx_state_t state_rx = gnrc_netdev->rx.state;
+
+    switch (state_rx) {
+        case RX_STATE_STOPPED: {
+            lwmac_rx_start(gnrc_netdev);
+            lwmac_rx_update(gnrc_netdev);
+            break;
+        }
+        case RX_STATE_FAILED: {
+            /* This may happen frequently because we'll receive WA from
```

Comment:
move to new function `static void _rx_management_failed(gnrc_netdev_t *gnrc_netdev);`?